plugins/hls-underlying-type-plugin/src/Ide/Plugin/UnderlyingType.hs

```diff
@@ -0,0 +1,141 @@
+      "Created action for" <+> pretty (printOutputable varName) <+> "->" <+> pretty (printOutputable typeName)
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId =
+  (defaultPluginDescriptor plId "Generates actions for going to the underlying type's definition.")
+    { pluginHandlers = mkPluginHandler SMethod_TextDocumentCodeAction (provider recorder)
+    }
+
+provider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_TextDocumentCodeAction
+provider recorder state _ (CodeActionParams _ _ (TextDocumentIdentifier uri) range _) = do
+  logWith recorder Debug $ LogPluginCalled uri range
+
+  nfp <- getNormalizedFilePathE uri
+
+  (HAR _ hieAst _ _ hieKind, posMapping) <-
+    runActionE "GetHieAst" state $
+      useWithStaleE GetHieAst nfp
+
+  let position = fromMaybe (range ^. L.start) $ fromCurrentPosition posMapping (range ^. L.start)
+      namesAtPosition = mconcat $ pointCommand hieAst position (extractNamesAtPosition hieKind)
+
+  logWith recorder Debug $ LogIdentifiersFound position namesAtPosition
+
+  actions <- case namesAtPosition of
+    [] -> pure []
+    identifiers -> do
+      typeActions <- forM identifiers $ \bindingName -> do
+        locationForIdentifierType <- runActionE "TypeCheck" state $ do
+          shakeExtras <- lift getShakeExtras
+          result <- liftIO $ runIdeAction "Get Type Definition" shakeExtras $ getTypeDefinition nfp position
+          pure $ fromMaybe [] result
+
+        logWith recorder Debug $ LogTypeLocationsFound locationForIdentifierType
+
+        typeDefActions <- forM locationForIdentifierType $ \(loc, identifier) -> do
+          logWith recorder Debug $ LogProcessingLocation loc identifier
+
+          case identifier of
+            Left _moduleName -> pure Nothing
+            Right underlyingTypeName -> do
+              logWith recorder Debug $ LogActionCreated bindingName underlyingTypeName
+              pure $ Just $ createGoToTypeDefAction bindingName underlyingTypeName loc
+
+        pure $ catMaybes typeDefActions
+
+      pure $ mconcat typeActions
+
+  pure $ InL actions
+
+extractNamesAtPosition :: HieKind a -> HieAST a -> [Name]
+extractNamesAtPosition hieKind ast =
+  case hieKind of
+    HieFresh -> rights $ map fst $ M.toList $ nodeIdentifiers $ nodeInfo ast
+    HieFromDisk {} -> []
+
+createGoToTypeDefAction :: Name -> Name -> Location -> (Command |? CodeAction)
+createGoToTypeDefAction boundVarName underlyingTypeName loc = do
+  let defRange = loc ^. L.range
+  InR $
+    CodeAction
+      ("Go to definition of " <> printOutputable underlyingTypeName <> " (inferred from " <> printOutputable boundVarName <> "'s type)")
+      (Just $ CodeActionKind_Custom "GoToUnderlyingTypeDefinition")
+      Nothing
+      Nothing
+      Nothing
+      Nothing
+      ( Just $
+          Command
+            "Go to definition"
+            -- TODO: How to decouple this from VS code?
```

Comment:
Introduce a `pluginCommands` handler, see the `PluginTutorial`, or `hls-cabal-plugin` for examples :)

Reply:
Thanks!